### PR TITLE
[FEA] Add third-party library integration testing of cudf.pandas to cudf

### LINKED ISSUE #16580: [FEA] Add integration testing of cudf.pandas
**Is your feature request related to a problem? Please describe.**
Currently we run tests of cudf.pandas against a number of third-party libraries to ensure that integrations work as expected. However, these tests are run independently of this repository and are not tracked regularly in CI. As a result, there is a possibility for these tests to get out of sync or start failing without anyone noticing.

**Describe the solution you'd like**
We should add integration tests into cudf's CI. These tests can be conda-only or pip-only to keep things simple. The tests would create a separate environment for each dependency of interest, pull in all necessary dependencies, and then run some validation. These tests could be conditioned upon most other Python tests passing to reduce overhead.

**Describe alternatives you've considered**
None

**Additional context**
In the immediate term we should probably just add these as nightly tests to minimize the impact on CI usage. If we later deem it necessary we can increase the frequency to per-PR. We ought to be able to structure the tests in such a way as to minimize CI usage (e.g. by using a single CI job that creates separate conda environments for each test).

### PR DESCRIPTION
**Pull request by Matt711**
## Description

Closes #16580 
## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

**Review by Matt711 — COMMENTED** [review]
I think I've migrated the files we need from [cudf-pandas-integration](https://github.com/rapidsai/cudf-pandas-integration) (ie. the tests and some of the shell scripts). And I've combined the three jobs before in to one.

I'm not sure if there's a proper way to test these changes, so I'll open the PR for review and get some feedback as it is.

File: .github/workflows/nightly.yaml
Comment (vyasr) [inline .github/workflows/nightly.yaml:1]: We already run nightly jobs from [test.yaml](https://github.com/rapidsai/cudf/blob/branch-24.10/.github/workflows/test.yaml) so you can merge this in there.

File: python/cudf/cudf_pandas_tests/third_party_integration_tests/ci/ci_run_library_tests.sh [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    -p cudf.pandas \
+    -v \
+    --continue-on-collection-errors \
+    --cache-clear \
+    --junitxml="${RAPIDS_TESTS_DIR}/junit-${lib}-cudf-pandas.xml" \
+    --numprocesses=${NUM_PROCESSES} \
+    --dist=worksteal \
+    ${TEST_DIR}/test_${lib}*.py \
+    ${test_keys}
+}
```
Comment (vyasr) [inline python/cudf/cudf_pandas_tests/third_party_integration_tests/ci/ci_run_library_tests.sh:41]: Minor nit: could we combine these functions using a boolean parameter to indicate whether or not to run with cudf.pandas? If you find it challenging, don't worry about it. We can always revisit in a follow-up.

File: python/cudf/cudf_pandas_tests/third_party_integration_tests/ci/ci_run_library_tests.sh
Comment (vyasr) [inline python/cudf/cudf_pandas_tests/third_party_integration_tests/ci/ci_run_library_tests.sh:1]: We can just put this in the top-level ci directory.

File: python/cudf/cudf_pandas_tests/third_party_integration_tests/ci/ci_run_library_tests.sh [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    --compare \
+    -p cudf.pandas \
+    -v \
+    --continue-on-collection-errors \
+    --cache-clear \
+    --junitxml="${RAPIDS_TESTS_DIR}/junit-${lib}-assertion.xml" \
+    --numprocesses=${NUM_PROCESSES} \
+    --dist=worksteal \
+    ${TEST_DIR}/test_${lib}*.py \
+    ${test_keys}
```
Comment (vyasr) [inline python/cudf/cudf_pandas_tests/third_party_integration_tests/ci/ci_run_library_tests.sh:62]: Do we actually need all these parameters for the comparison phase? I suspect that they are all superfluous.

File: python/cudf/cudf_pandas_tests/third_party_integration_tests/ci/extract_lib.sh
Comment (vyasr) [inline python/cudf/cudf_pandas_tests/third_party_integration_tests/ci/extract_lib.sh:1]: This is a small enough file, at this point I'd say we should just combine this into the test.sh script and run a single CI step. That also matches our other CI jobs more closely (one script per job).

**Review by vyasr — CHANGES_REQUESTED** [review]
I've recommended moving the test from nightly.yaml to test.yaml, but for the lifetime of this PR can we also add it to pr.yaml? That way we can test it out during this PR and iron out any issues, then remove that before the end so that the test only runs nightly.

File: .github/workflows/test.yaml [hunk 3]
```diff
@@ -124,3 +124,22 @@ jobs:
       date: ${{ inputs.date }}
       sha: ${{ inputs.sha }}
       script: ci/cudf_pandas_scripts/run_tests.sh
+  integration-tests:
```
Comment (Matt711) [inline .github/workflows/test.yaml:127]: Change this name to match `pr.yaml`

File: ci/cudf_pandas_scripts/run_tests.sh [hunk 4]
```diff
@@ -62,6 +62,7 @@ else
 fi
 
 python -m pytest -p cudf.pandas \
+    --ignore=./python/cudf/cudf_pandas_tests/third_party_integration_tests/ \
```
Comment (Matt711) [inline ci/cudf_pandas_scripts/run_tests.sh:65]: I think we want to keep all `cudf.pandas` test under one directory. So this line makes the `unit-test-cudf-pandas` job ignores the integration tests subdirectory.

File: .github/workflows/pr.yaml [hunk 5]
```diff
@@ -296,3 +297,14 @@ jobs:
         node_type: cpu4
         build_type: pull-request
         run_script: "ci/cudf_pandas_scripts/pandas-tests/diff.sh"
+  third-party-integration-tests-cudf-pandas:
```
Comment (Matt711) [inline .github/workflows/pr.yaml:300]: TODO: Will remove when this PR is approved

File: python/cudf/cudf_pandas_tests/third_party_integration_tests/ci/ci_run_library_tests.sh (hunk 2 above)
Comment (Matt711) [reply python/cudf/cudf_pandas_tests/third_party_integration_tests/ci/ci_run_library_tests.sh]: Some of them like dist, numprocesses are used in other ci scripts, but I removed the others

File: python/cudf/cudf_pandas_tests/third_party_integration_tests/ci/ci_run_library_tests.sh (hunk 1 above)
Comment (Matt711) [reply python/cudf/cudf_pandas_tests/third_party_integration_tests/ci/ci_run_library_tests.sh]: Okay I’ll follow-up with a PR that combines these two functions

File: ci/cudf_pandas_scripts/third-party-integration/ci_run_library_tests.sh
Comment (bdice) [inline ci/cudf_pandas_scripts/third-party-integration/ci_run_library_tests.sh:1]: Most of the directories are named with underscores but this one has dashes. Can we rename this like `ci/cudf_pandas_scripts/third_party_integration/...`?

Hmm. I also see we have `pandas-tests` in the same directory. Maybe it should be renamed too, in a later PR.

File: python/cudf/cudf_pandas_tests/third_party_integration_tests/dependencies.yaml [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+        packages:
+          - pre-commit
+  cuda_version:
+    specific:
+      - output_types: conda
+        matrices:
+          - matrix:
+              cuda: "11"
+            packages:
+              - cuda-version=11.8
```
Comment (bdice) [inline python/cudf/cudf_pandas_tests/third_party_integration_tests/dependencies.yaml:141]: Do we need both `cuda: "11"` and `cuda: "11.8"`? We typically do not list major-only versions in our dependencies files.

**Review by bdice — APPROVED** [review]
I didn't read the tests that closely since we have previously reviewed them. I focused on the integration pieces. This seems fine to me! Thanks for working on this.

File: python/cudf/cudf_pandas_tests/third_party_integration_tests/dependencies.yaml (hunk 6 above)
Comment (Matt711) [reply python/cudf/cudf_pandas_tests/third_party_integration_tests/dependencies.yaml:141]: I got errors like "Cannot find matching value in matrix with cuda:11/12", so I added both.

File: ci/cudf_pandas_scripts/third-party-integration/ci_run_library_tests.sh
Comment (Matt711) [reply ci/cudf_pandas_scripts/third-party-integration/ci_run_library_tests.sh:1]: Thanks! Yes, address these in a follow up

**Review by vyasr — APPROVED** [review]
LGTM. I didn't re-review the actual tests since those have been looked at before, just the CI/GHA scripts and dependencies.yaml.

File: .github/workflows/pr.yaml [hunk 7]
```diff
@@ -296,3 +297,14 @@ jobs:
...
+    needs: wheel-build-cudf
```
[the remark this answers is not in the document]
Comment (Matt711) [reply .github/workflows/pr.yaml:301]: Ah got it, thanks for the note

File: ci/cudf_pandas_scripts/run_tests.sh (hunk 4 above)
Comment (Matt711) [reply ci/cudf_pandas_scripts/run_tests.sh:68]: I think they we're lost when I need to revert some other things I was trying. I'll add the comment back.

**Pull request merged**